### DISCUSSION TITLE
BLOCKER : Converting/Repurposing UWP Shared Project in WinUI3 project (XamlParseException / NavigationView)

### DISCUSSION THREAD

**Discussion by ToddGlodek** [unoplatform/uno#8880] [Q&A] [answered]
Last year - I wrote a proof of concept app using Uno for UWP.   Now that Uno's WinUI support I thought I should upgrade it.   I created a new project/solution  using the latest ( 4.3.6.9) templates.   I thought, "Since a shared project is just a list of files"  I ought to be able to just replace the newly created one with the one from my UWP project.   I then changed all the occurrences of  'Windows.UI' to 'Microsoft.UI'  and made a few corrections to  get it to compile.  Getting it to compile was generally straight forward.
 
 However, when I try to run the app, I get a Microsoft.UI.Xaml.Markup.XamlParseException: 'XAML parsing failed'  during InitializeComponent() when it tries to load the MainPage in the updated shared project.     The only content on the MainPage is a NavigationView.  When I remove all references to the NavigationView control in the  XAML - it runs fine (displays 'Hello World').    
 
 So I thought - ' maybe I have to rebuild the NavigationView layout from scratch for WinUI ?? '    However,  just adding a stub reference to the NavigationView results in the XamlParseException
 
 ```
<Page
    x:Class="UnoTest2.Shared.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" 
    mc:Ignorable="d"
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}"  Loaded="OnLoad">

    <NavigationView  />

</Page>
```

However,  If I  copy/paste the MainPage.Xaml  from old (converted) Shared Project into the (new) Shared Project that was just created for the WinUI3 solution - then it displays the Navigation control with no problems.    Since the XAML displays  fine in one Share Project - but not in the other - I'm concluding that it is not actually a XAML parsing problem.  

I didnt notice anything in [Platform-specific XAML markup in Uno]( https://platform.uno/docs/articles/platform-specific-xaml.html) that was an obvious problem.  Besides, I am using Windows for both tests.

Both projects have the same values in the project file ( except for RootNamespace )

`<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>WinExe</OutputType>
		<TargetFramework>net6.0-windows10.0.19041.0</TargetFramework>
		<TargetPlatformMinVersion>10.0.18362.0</TargetPlatformMinVersion>
		<RootNamespace>UnoTestWinUI3</RootNamespace>
		<ApplicationManifest>app.manifest</ApplicationManifest>
		<Platforms>x86;x64;arm64</Platforms>
		<RuntimeIdentifiers>win10-x86;win10-x64;win10-arm64</RuntimeIdentifiers>
		<PublishProfile>win10-$(Platform).pubxml</PublishProfile>
		<UseWinUI>true</UseWinUI>
		<EnablePreviewMsixTooling>true</EnablePreviewMsixTooling>
		<StartupObject></StartupObject>
	</PropertyGroup>
`

Does it matter that the old (converted) Shared Project files are all in a totally different directory unrelated to the new WinUI project?   When I look at the  MainPage.g.i.cs - it seems to have generated all the appropriate code.   NOTE, the /obj/  directory containing MainPage.g.i.cs also looks like it has code generated files associated with all the pages from the old (converted) Shared Project files.  It looks like it is finding all the files, but I'm not an expert in  what .g.i.cs  files should look like.

Any Idea waht is going on here ??   Can you point me to some documentation related to UPGRADING FROM OLDER VERSIONS OF UNO PROJECTS ??

**Comment by jeromelaban** [comment, marked as answer]
It's hard to tell what the issue is here, but in general the best is to import incrementally to find out where something breaks. Add pages and controls one by one, and you'll be able to isolate what's breaking.

**Reply by ToddGlodek** [reply]
I appreciate your quick reply on this - even though it is totally the opposite of what I wanted to hear.

**Reply by jeromelaban** [reply]
Something you can try also is to run the WinAppSDK head using the native+managed debugger (in the debug options of the project), sometimes it gives additional insights.

**Reply by ToddGlodek** [reply]
Im not sure - but I thing it might be related to the way the control library is created ??

```
  <ItemGroup Condition="'$(TargetFramework)'=='net5.0-windows10.0.18362'">   **<<    Says NET5.0  seems like it should say 6.0**   
    <PackageReference Include="Microsoft.WindowsAppSDK" Version="1.0.0" />
    <FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" RuntimeFrameworkVersion="10.0.22000.24" />
    <FrameworkReference Update="Microsoft.Windows.SDK.NET.Ref" TargetingPackVersion="10.0.22000.24" />
  </ItemGroup>
```

**Reply by jeromelaban** [reply]
Indeed it should be net6.0 instead. Where did you find this code? We may need to adjust it in samples.

**Reply by jeromelaban** [reply]
Thanks, this is indeed incorrect. https://github.com/unoplatform/uno/pull/8898 will fix this, but you can change for `net6.0-windows10.0.18362` instead.

**Comment by ToddGlodek** [comment]
It DID  come from a the NEW LIBRARY project template

Sent from Mail<https://go.microsoft.com/fwlink/?LinkId=550986> for Windows

**Comment by ToddGlodek** [comment]
Troubleshooting build issues : https://platform.uno/docs/articles/uno-builds-troubleshooting.html#generating-msbuild-binary-log-files

**Comment by ToddGlodek** [comment]
@jeromelaban ~ per your instruction to "Add pages and controls one by one, and you'll be able to isolate what's breaking"...   I have isolated one XAML based UserControl that is still giving me problems.  

As part of the conversion, changed all the occurrences of 'Windows.UI' to 'Microsoft.UI'.  This, however, is giving me issues because I get "The type or namespace name 'Color' does not exist in the namespace 'Microsoft.UI' (are you missing an assembly reference?)"  I cannot find any online documentation that refers to Microsoft.UI.Color (only to Windows.UI.Color).  I am noticing that even in the Microsoft documentation for WinUI controls (such as [color picker](https://docs.microsoft.com/en-us/windows/winui/api/microsoft.ui.xaml.controls.colorpicker.color?view=winui-2.7#microsoft-ui-xaml-controls-colorpicker-color) )  - they continue to use Windows.UI.Color.  Also, when you look at Microsoft.UI.Colors - each individual named color is of type Windows.UI.Color.

So - I thought "maybe in this particular case I should use Windows.UI.Color".  However, when I reference Windows.UI.Color, I begin to get the "Index (zero based) must be greater than or equal to zero and less than the size of the argument list."  which I reported back in Februrary ([BUILD FAILS reports issue in Microsoft.Windows.UI.Xaml.Common.targets](https://github.com/unoplatform/uno/discussions/8145)). 



In that case you asked me to "provide a binlog" - but instead I thought it would be better to upgrade to WinUI instead.  But now - it looks like I am right back at the same spot.  Can you think of something quick and easy I should do to resolve the Color issue - or should I give you the binlog as you originally requested ?

**Reply by jeromelaban** [reply]
Thanks. Yes, a binlog will help.

**Comment by ToddGlodek** [comment]
When I ran dotnet build  at the command prompt - I got this a output.  It looks like it  is looking for a net50 nuget package related to XamlCompiler.  I thought that was odd since I am using net6 only.   I deleted the nuget package thinking it might update it to a net6 version – but it just reinstalled the same net5 nuget.


> dotnet build /bl .\XXXXX.Controls.csproj
Microsoft (R) Build Engine version 17.2.0+41abc5629 for .NET
Copyright (C) Microsoft Corporation. All rights reserved.

C:\Program Files\dotnet\sdk\6.0.300\MSBuild.dll -consoleloggerparameters:Summary -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,C:\Program Files\dotnet\sdk\6.0.300\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,C:\Program Files\dotnet\sdk\6.0.300\dotnet.dll -maxcpucount -restore -verbosity:m /bl .\XXXXX.Controls.csproj
  Determining projects to restore...
  Restored C:\Users\toddg\source\repos\XXXXXXWinUI3\XXXXXX.Controls\XXXXXX.Controls.csproj (in 559 ms).
  1 of 2 projects are up-to-date for restore.
  XXXXX.Core -> C:\Users\t____\source\repos\XXXXXXWinUI3\XXXXXX.core\bin\Debug\net6.0\XXXXX.Core.dll
C:\Users\t____\.nuget\packages\microsoft.windowsappsdk\1.0.3\buildTransitive\Microsoft.UI.Xaml.Markup.Compiler.interop.
targets(558,9): error MSB3073: The command ""C:\Users\t____\.nuget\packages\microsoft.windowsappsdk\1.0.3\buildTransiti
ve\..\tools\**net5.0**\..\**net472**\XamlCompiler.exe" "obj\Debug\net6.0-windows10.0.18362\\input.json" "obj\Debug\net6.0-windo
ws10.0.18362\\output.json"" exited with code 1. [C:\Users\t__\source\repos\XXXXXWinUI3\XXXXX.Controls\XXXXXX.Controls.csproj]

Build FAILED.

C:\Users\t____\.nuget\packages\microsoft.windowsappsdk\1.0.3\buildTransitive\Microsoft.UI.Xaml.Markup.Compiler.interop.
targets(558,9): error MSB3073: The command ""C:\Users\t____\.nuget\packages\microsoft.windowsappsdk\1.0.3\buildTransiti
ve\..\tools\net5.0\..\net472\XamlCompiler.exe" "obj\Debug\net6.0-windows10.0.18362\\input.json" "obj\Debug\net6.0-windo
ws10.0.18362\\output.json"" exited with code 1. [C:\Users\toddg\source\repos\XXXXXXWinUI3\Prognosti.Controls\XXXXXX.Controls.csproj]
    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:05.21



Subject: Re: [unoplatform/uno] BLOCKER : Converting/Repurposing UWP Shared Project in WinUI3 project (XamlParseException / NavigationView) (Discussion #8880)

**Reply by jeromelaban** [reply]
The version is likely unrelated to the issue here. This is not something Uno managed as it is part of the WinUI/WinAppSDK toolchain. If you're able to isolate this issue in repro project, you can post this to the WinAppSDK or WinUI repositories.
